src/nvim/ex_docmd.c

```diff
@@ -9504,6 +9504,13 @@ static void ex_terminal(exarg_T *eap)
   }
 
   do_cmdline_cmd(ex_cmd);
+
+  if (*eap->arg == NUL) {
+    do_cmdline_cmd("autocmd TermClose <buffer> "
+                   " if !v:event.status | "
+                   "   exec 'bdelete! ' .. expand('<abuf>') | "
+                   " endif");
```

Comment:
Probably should be global rather than `<buffer>` local. And we can declare it in a `nvim_` augroup so that users can opt-out.

Reply:
I don’t think it should be global as then it would apply to all `termopen` calls rather than just `:terminal`. The former provides finer grained control so we should make fewer assumptions and allow the user to do whatever they want, while `:terminal` is a convenience function that should have more convenient defaults.

My 2¢.

Reply:
Hmm. Thinking about how a user could opt-out of this in its current form: `autocmd TermClose` is done after the `do_cmdline_cmd(ex_cmd);` above. So this is setup after TermOpen... how would users opt-out?

> I don’t think it should be global as then it would apply to all `termopen` calls

Makes sense, but OTOH consistent behavior is easier to reason about, and to document. Clear documentation (one-liner that can be pasted), somewhat addresses the termopen() case, and avoids contortions for the :terminal case.

A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts:  can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

Reply:
> A global autocmd or augroup solves the user config case, but doesn't solve the termopen() case for scripts: can't selectively opt-out, except by `:noautocmd call termopen(...)` ... maybe that's reasonable...

I think that's reasonable, especially if it's documented.

If we do make it a global autocommand then we may want to resolve https://github.com/neovim/neovim/pull/15401 first so that we can use whatever default autocommand framework we decide on.

Reply:
I've been thinking about this more and two things occurred to me:

1) Making it a global autocommand means we can't use the heuristic of whether or not an argument was provided to `:terminal` to determine if we should automatically close the buffer or not. Right now, we create this autocommand if there is no argument and we are therefore using a default shell. But if, for instance, a user runs `:terminal ls` then we *do not* want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

2) I wonder if it's not actually more common that users may want to opt-out of the auto-close behavior on a per-job/per-terminal instance, rather than opting out globally. This also provides plugins finer grained control to opt out of this behavior, which wouldn't be possible (or at least, more complex) with a global autocommand.

For these two reasons I think we ought to stick with a buffer-local autocommand. What do you think?

Reply:
> 1\. ... But if, for instance, a user runs `:terminal ls` then we _do not_ want to automatically close the buffer when the job completes. If we use a global autocommand we will need another flag or something to indicate which state we are in.

yeah, that's where I was going with [this comment](https://github.com/neovim/neovim/pull/15440#discussion_r693233441):

> In a global TermOpen autocmd, that could possibly be derived from the `term://` buffer name, or maybe we should expose the args in `v:event`. Or `nvim_get_chan_info(&channel)`

re (2):

> This is much more straightforward to do with a buffer-local autocommand. This would also provide plugins finer-grained control to opt out of this behavior,

Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`

Reply:
> Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

The plugin would need to disable the global autocommand in that case though, which the user may not want. The only way around this would be to add some kind of buffer-local variable that acts as a flag (e.g. `let b:terminal_autoclose = v:false`) that the plugin could set. But if we're doing that, why not just use a buffer-local autocommand?

I guess I disagree that a global autocommand is less messy, it paints with too broad of a brush so we end up needing to introduce further measures to account for edge cases and exceptions.

Reply:
>> Should work, since this default autocmd will always be defined before any user/plugin autocmds. Just needs a unique augroup, something like `nvim_terminal_close`.

^ this was regarding buffer-local autocommand

Reply:
Ah cool, sorry for the misunderstanding 😅